intellij-plugin-structure/structure-hub/src/main/kotlin/com/jetbrains/plugin/structure/hub/HubPluginManager.kt

```diff
@@ -43,46 +46,82 @@ class HubPluginManager private constructor() : PluginManager<HubPlugin> {
+    val manifestContent = manifestFile.readText()
+    val manifest = Json(
+        JsonConfiguration.Stable.copy(isLenient = true, ignoreUnknownKeys = true)
+    ).parse(HubPluginManifest.serializer(), manifestContent)
+    val icon = try {
+      loadIcon(pluginDirectory, manifest.iconUrl)
+    } catch (e: IllegalArgumentException) {
+      return PluginCreationFail(HubIconInvalidUrl())
+    }
+    return createPlugin(manifest, manifestContent, icon)
   }
 
-  private fun loadDescriptor(descriptorFile: File): PluginCreationResult<HubPlugin> {
+  private fun createPlugin(
+      descriptor: HubPluginManifest, manifestContent: String, icon: PluginIcon?
+  ): PluginCreationResult<HubPlugin> {
     try {
-      val manifestContent = descriptorFile.readText()
-      val descriptor = Json(JsonConfiguration.Stable.copy(isLenient = true, ignoreUnknownKeys = true)).parse(HubPlugin.serializer(), manifestContent)
-      descriptor.manifestContent = manifestContent
-      val vendorInfo = parseHubVendorInfo(descriptor.author)
-      descriptor.apply {
-        vendor = vendorInfo.vendor
-        vendorEmail = vendorInfo.vendorEmail
-        vendorUrl = vendorInfo.vendorUrl
-      }
       val beanValidationResult = validateHubPluginBean(descriptor)
       if (beanValidationResult.any { it.level == PluginProblem.Level.ERROR }) {
         return PluginCreationFail(beanValidationResult)
       }
-      return PluginCreationSuccess(descriptor, beanValidationResult)
+
+      val vendorInfo = parseHubVendorInfo(descriptor.author ?: "")
+      if (vendorInfo.vendor == null) return PluginCreationFail(PropertyNotSpecified("author"))
+      val plugin = with(descriptor) {
+        HubPlugin(
+            pluginId = pluginId,
+            pluginName = pluginName,
+            pluginVersion = pluginVersion,
+            url = url,
+            description = description,
+            dependencies = dependencies,
+            products = products,
+            vendor = vendorInfo.vendor,
+            vendorEmail = vendorInfo.vendorEmail,
+            vendorUrl = vendorInfo.vendorUrl,
+            manifestContent = manifestContent,
+            icons = if (icon != null) listOf(icon) else emptyList()
+        )
+      }
+      return PluginCreationSuccess(plugin, beanValidationResult)
     } catch (e: Exception) {
       e.rethrowIfInterrupted()
       LOG.info("Unable to read plugin descriptor $DESCRIPTOR_NAME", e)
       return PluginCreationFail(UnableToReadDescriptor(DESCRIPTOR_NAME, e.localizedMessage))
     }
   }
+
+  private fun loadIcon(pluginDirectory: File, iconUrl: String?): PluginIcon? {
+    return when {
+      iconUrl == null -> null
+      iconUrl.contains("://") || iconUrl.startsWith("/") -> {
+        LOG.warn("Unsupported widget iconUrl: '$iconUrl'")
+        null
+      }
+      else -> {
+        val iconEntry = File(pluginDirectory, iconUrl)
+        if (!iconEntry.exists()) throw IllegalArgumentException("Icon file doesn't exist")
```

Comment:
use `require { }`

Reply:
and why IllegalArgumentException? We use PluginCreationFail over here

Reply:
but still keep doing this above
```kotlin
    } catch (e: IllegalArgumentException) {
      return PluginCreationFail(HubIconInvalidUrl())
    }
```

Reply:
And what am I supposed to do with PluginCreationFail in method that returns icon?

Reply:
<img width="303" alt="Screen Shot 2020-03-24 at 11 17 04 AM" src="https://user-images.githubusercontent.com/10503748/77403255-05b3b300-6dc1-11ea-8bfd-31bc87b63e7c.png">
okay, but you dont need the message `"Icon file doesn't exist`

Reply:
just use require